4029 Fix V4 Search API Serializer errors and other bugs

albertisfu opened the pull request:
This PR fixes the V4 Search API serializer errors described in #4029:

- The first error relates to the **`cites`** field. The issue was that this field was added after the initial indexing, so some `RECAPDocuments` in the index still lack it. The problem was only present in the **`RECAP_DOCUMENT`** search type. The DRF field serializer uses a **`getattr`** method, which lacks a default value; thus, when a field is missing in the document, a `KeyError` is raised. When rendering nested child documents, we use the **`ESResultObject`** to wrap each child's result data so that data from child documents can be properly retrieved according to their data type. This class overrides the **`getattr`** method and adds a default value of `None` for missed fields, which is why the `RECAP` type doesn't raise the error. Thus, the solution was to also wrap parent documents in an `ESResultObject`.

  The remaining issue is that a document with no **`cites`** field will be rendered as **`null`** instead of an empty list. To solve this, I used the **`NoneToListField`** for the **`cites`** field (which is the same workaround I applied in #4021 for the ES DSL bug) so that **`None`** fields are rendered as empty lists.

- The other serializer error concerned the **`pacer_case_id`** field, which in the DB is a **`CharNullField`**. When this value is `null`, it returns an empty string instead of **`None`**. The **`pacer_case_id`** was being cast to an **`IntegerField`**, which doesn't accept empty strings. The solution was to avoid the **`IntegerField`** and instead render the **`pacer_case_id`** as a `CharField` in the API response.
- Additionally, the docket count bug described in #4032 has been fixed. The solution was simple: the comparison to display the approximate count should be `< ELASTICSEARCH_MAX_RESULT_COUNT` instead of `<= ELASTICSEARCH_MAX_RESULT_COUNT`.
- I also identified a potential issue during testing that could occur in production. When HL is disabled, the `plain_text` snippet is extracted from the DB. However, there could be an error rendering the page in the case of a race condition, after the results on a page are returned by ES and before the `plain_text` snippet is extracted from the DB. If a document belonging to the page is removed from the DB, a `KeyError` is raised. To fix this, if the document is not found in the DB, the `plain_text` related to the document will be returned as empty, allowing the correct rendering of the page.

For the other issues in #4032, I will open a separate PR as I continue debugging the sorting issue.

mlissner reviewed (approved):
Looks ok to me at a skim, but big enough to deserve a better review than I can do lately, so I'll put it on Eduardo's backlog so he can take a look too. But feel free to merge when y'all are happy.

ERosendo commented on cl/search/api_utils.py:
```diff
@@ -36,6 +36,23 @@
+        return self._data
+
+
+class ESResultObject(ResultObject):
+
+    def __getattr__(self, key):
+        return getattr(self._data, key, None)
```
Instead of this custom class, let's use the built-in `defaultdict` class. This will simplify the code and provides automatic default values for missing attributes or keys when accessed.

ERosendo reviewed (approved):
The code looks solid! Let's merge it after addressing my comment 👍

albertisfu commented:
@ERosendo Thanks! I've applied your suggestion and, since `ESResultObject` is no longer required, I removed it. During the process, I also removed some dead code in the V3 `ESList`, which I had forgotten to remove when working on an initial V4 approach.
